fix(transformer): improve legacy decorator handling and fix constructor parameter decorators

Dunqing opened the pull request:
Fixes https://github.com/rolldown/rolldown/issues/6079. Regressed introduced by https://github.com/oxc-project/oxc/pull/13227.


The root cause is that the class plugin will insert a constructor method in some cases, but metadata handling is processed before the constructor is inserted, which causes the method count not to be exactly the same as the method metadata count. Thus, decorator metadata is inserted in the incorrect place.

This PR refactors the core transformation logic to ensure that the new methods and properties inserted by other plugins won't affect the decorator transformation.

### Transformation Flow Changes

  Before: Complex upfront analysis with separate processing paths
  - Used check_class_has_decorated() to analyze entire class tree
  - Separate handling for class vs element decorators
  - Multiple disconnected stacks for state tracking

  After: Unified incremental processing
  - Process decorators as we traverse (exit_method_definition,
  exit_property_definition, etc.)
  - Accumulate decoration statements in single stack during traversal
  - Unified processing in transform_class() using consolidated state

Dunqing commented:
* **#13632** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13632?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13632?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13632).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_transformer/src/decorator/legacy/mod.rs:
```diff
@@ -104,8 +115,9 @@ impl<'a, 'ctx> LegacyDecorator<'a, 'ctx> {
             metadata: LegacyDecoratorMetadata::new(ctx),
             class_decorated_data: None,
-            ctx,
             decorations: FxHashMap::default(),
+            class_decoration_stack: NonEmptyStack::new(ClassDecorations::default()),
```
The order is correct, please double-check your review.

overlookmotel commented on crates/oxc_transformer/src/decorator/legacy/mod.rs:
```diff
@@ -326,22 +450,24 @@ impl<'a> LegacyDecorator<'a, '_> {
+            return;
+        }
 
+        let current_class_decorations = self.class_decoration_stack.pop();
```
`class_decoration_stack` is always pushed to in `enter_class` (regardless of whether it's a class expression, or whether `declare == true`).

So I think the `pop` also needs to happen unconditionally - before the `if class.is_expression() || class.declare` check.

Otherwise, the stack will get out of sync with something like:

```js
class C {
  @dec method1() {}
  method2() {
    return class {}; // class expression
  }
  @dec method3() {}
}
```

When using stacks, I think it's generally a good idea to `debug_assert!` that the stack has been emptied again in `exit_program`. This commit adds such a check, and it panics when running conformance: https://github.com/oxc-project/oxc/commit/ae084fa932c43e13e5cae2fd36a9e4d280762ddb

I suggest adding similar checks in `exit_program` for `method_metadata_stack` and `constructor_metadata_stack` too, to make sure they're operating as intended.

overlookmotel commented on crates/oxc_transformer/src/decorator/legacy/mod.rs:
```diff
@@ -81,6 +72,22 @@ struct ClassDecoratedData<'a> {
     alias_binding: Option<BoundIdentifier<'a>>,
 }
 
+/// Decoration state for a single class level.
+/// This struct groups together all the state needed while processing decorators for one class,
+/// allowing clean management of nested class scenarios through a stack of these structs.
```
AI wrote this comment, right? You can tell by how it's excessively long, and the last 2 lines don't really say anything! 😄

```suggestion
/// Decoration state for a single class.
```

Dunqing replied:
Ah, yes. Most of the comments are written by AI, and I tweak them a little bit.

overlookmotel reviewed (commented):
This PR changes quite a lot, and I'm not that familiar with decorators transform to start with, so it's hard to review properly without spending hours on it.

There's one bug related to stacks (see below).

Also, a couple of comments/questions:

1. Are you sure that decorators always get transformed by other transforms before they're removed from AST? (e.g. in a case like `@(x ?? y)`)

2. How `method_metadata_stack` and `constructor_metadata_stack` are pushed/popped looks a little fragile. Is it guaranteed that every `push` has a corresponding `pop`, even if some other transform e.g. adds/removes a constructor or method in between, or if some classes are class expressions?

This decorators transform is pretty damn tricky!

(I hope you don't mind I pushed 1 unimportant commit, just to reformat the test cases. Conformance CI fail is not related to that commit - it passes locally. Something must have changed on main since PR was originally submitted.)

overlookmotel reviewed (approved):
I added a commit just to rename a var to be more accurate description of what it is.

One nit comment below. If you're willing to make that change, please go ahead and merge afterwards. Or if you can't be bothered, just merge it now!

Dunqing committed:
fix(transformer): improve legacy decorator handling and fix constructor parameter decorators (#13632)

Fixes https://github.com/rolldown/rolldown/issues/6079. Regressed introduced by https://github.com/oxc-project/oxc/pull/13227.

The root cause is that the class plugin will insert a constructor method in some cases, but metadata handling is processed before the constructor is inserted, which causes the method count not to be exactly the same as the method metadata count. Thus, decorator metadata is inserted in the incorrect place.

This PR refactors the core transformation logic to ensure that the new methods and properties inserted by other plugins won't affect the decorator transformation.

### Transformation Flow Changes

  Before: Complex upfront analysis with separate processing paths
  - Used check_class_has_decorated() to analyze entire class tree
  - Separate handling for class vs element decorators
  - Multiple disconnected stacks for state tracking

  After: Unified incremental processing
  - Process decorators as we traverse (exit_method_definition,
  exit_property_definition, etc.)
  - Accumulate decoration statements in single stack during traversal
  - Unified processing in transform_class() using consolidated state
